src/OrchardCore.Modules/OrchardCore.AdminDashboard/Drivers/DashboardContentDisplayDriver.cs

```diff
@@ -0,0 +1,109 @@
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Http;
+using OrchardCore.ContentManagement;
+using OrchardCore.ContentManagement.Display.ContentDisplay;
+using OrchardCore.ContentManagement.Display.ViewModels;
+using OrchardCore.Contents;
+using OrchardCore.DisplayManagement.Handlers;
+using OrchardCore.DisplayManagement.Views;
+
+
+namespace OrchardCore.AdminDashboard.Drivers
+{
+    public class DashboardContentDisplayDriver : ContentDisplayDriver
+    {
+        private readonly IHttpContextAccessor _httpContextAccessor;
+        private readonly IAuthorizationService _authorizationService;
+        private readonly IContentManager _contentManager;
+
+
+        public DashboardContentDisplayDriver(IHttpContextAccessor httpContextAccessor,
+            IAuthorizationService authorizationService,
+            IContentManager contentManager)
+        {
+            _httpContextAccessor = httpContextAccessor;
+            _authorizationService = authorizationService;
+            _contentManager = contentManager;
+
+        }
+
+        public override async Task<IDisplayResult> DisplayAsync(ContentItem model, BuildDisplayContext context)
+        {
+            var httpContext = _httpContextAccessor.HttpContext;
+            var dashboardFeature = httpContext.Features.Get<DashboardFeature>();
+
+            if (dashboardFeature == null || !dashboardFeature.IsManageRequest)
+            {
+                return null;
+            }
+
+            var results = new List<IDisplayResult>();
+
+            var hasPublished = await _contentManager.HasPublishedVersionAsync(model);
+            var hasDraft = model.HasDraft();
+            var hasEditPermission = await _authorizationService.AuthorizeAsync(httpContext.User, CommonPermissions.EditContent, model);
+            var hasDeletePermission = await _authorizationService.AuthorizeAsync(httpContext.User, CommonPermissions.DeleteContent, model);
+            var hasPublishPermission = await _authorizationService.AuthorizeAsync(httpContext.User, CommonPermissions.PublishContent, model);
+
+            var dragHandle = Dynamic("Dashboard_DragHandle", m =>
```

Comment:
```suggestion
            var dragHandle = Initialize<ContentItemViewModel>("Dashboard_DragHandle", m =>
```

Great thanks for moving it to a feature, and cleaning up the async.

Suggestion here is that we are trying to move away from dynamic where possible, and of all the shape factory creations, this is the better one for allocations. It will create a strongly typed `ContentItemViewModel` for each, which means you can use that as the model in the views, rather than dynamic.

I'm not seeing any reason for them to be dynamic, unless I am wrong?

So here, and all the other shapes from this driver, look like they can be initialized.

Can you also have a look through and clean up all the extra line breaks too please ?

Reply:
done